library/core/src/option.rs

```diff
@@ -689,6 +689,100 @@ impl<T> Option<T> {
         }
     }
 
+    const fn get_some_offset() -> isize {
+        if core::mem::size_of::<Option<T>>() == core::mem::size_of::<T>() {
+            // niche optimization means the `T` is always stored at the same position as the Option.
+            0
+        } else {
+            assert!(
+                core::mem::size_of::<Option<T>>()
+                    == core::mem::size_of::<Option<core::mem::MaybeUninit<T>>>()
+            );
+            let some_uninit = Some(core::mem::MaybeUninit::<T>::uninit());
+            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
+            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
+            // the same layout, which we have asserted above.
+            unsafe {
+                (some_uninit.as_ref().unwrap() as *const core::mem::MaybeUninit<T> as *const u8)
+                    .offset_from(
+                        &some_uninit as *const Option<core::mem::MaybeUninit<T>> as *const u8,
+                    )
+            }
+        }
+    }
+
+    /// Returns a slice of the contained value, if any. If this is `None`, an
+    /// empty slice is returned. This can be useful to have a single type of
+    /// iterator over an `Option` or slice.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(option_as_slice)]
+    ///
+    /// assert_eq!(
+    ///     [Some(42).as_slice(), None.as_slice()],
+    ///     [&[42][..], &[][..]],
+    /// );
+    /// ```
+    #[inline]
+    #[must_use]
+    #[unstable(feature = "option_as_slice", issue = "none")]
+    pub fn as_slice(&self) -> &[T] {
+        // if T is uninhabited, the value is always `None`.
+        if core::mem::size_of::<Option<T>>() == 0 {
+            return &[];
```

Comment:
I don't think this is the same as being uninhabited? I would expect `Some(()).as_slice()` to return `&[()]`

Reply:
Oh I see, this is if the *Option* has size 0, i.e. there's no discriminant. Maybe leave that as a comment.

Reply:
This special case is unnecessary anyways, it generates just as good code without it: https://godbolt.org/z/x66P4jasY.

(Because it's trivial for LLVM to see that `get_some_offset()` is 0 and `is_some()` is false for `Option<!>`)

Reply:
Uninhabited types don't have to be ZST at all, e.g. `(i32, !)` still has size 4.

Reply:
That's a type with an uninhabited member, which is different.

Reply:
> I probably should've noticed your credentials and not expressed my initial thought.

Don't worry about that -- I'm allowed to be wrong too. 🙂 I wasn't sure whether `Option<(i32, !)>` would collapse, so thanks for checking that it doesn't. (At least not yet...)

Reply:
Ok, that's likely the nail in the coffin for the uninhabitedness check via Option size.

Anyway, since we can ignore this anyway, I'll simply remove the check.